network/codec/json/codec.go

```diff
@@ -75,3 +75,20 @@ func (c *Codec) Decode(data []byte) (interface{}, error) {
 
 	return v, nil
 }
+
+// DecodeMsgType is a helper func that returns the underlying message of the message
+func (c *Codec) DecodeMsgType(data []byte) (byte, string, error) {
```

Comment:
The idea of decoupling the message type decoding is nice. Though, the `JSON` implementation of it requires the entire message to be decoded, right? Then by switching to `JSON`, we are silently incurring the overhead of additional decoding per message. 

Is it possible to also encode `JSON` similar to `CBOR` by appending the first byte as the message type (and potentially getting rid of the `Envelope`)?